Enable ufw and update allowed port rules

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR does the following things:
* Enables `ufw` (previously, we left `ufw` disabled)
* Modifies the list of ports allowed by `ufw` to those specified in https://github.com/cisagov/cool-system-internal/issues/80
 
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

These changes were requested by our Assessments team in order to make these instances more secure when they are initially deployed.  This work is part of https://github.com/cisagov/cool-system-internal/issues/80.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I deployed a Kali instance with these changes to a Staging environment and confirmed that `ufw` was active when the instance started up and the expected rules were set:

```
$ sudo ufw status numbered
Status: active

     To                         Action      From
     --                         ------      ----
[ 1] 22/tcp                     ALLOW IN    Anywhere                  
[ 2] 5000:5999/tcp              ALLOW IN    Anywhere                  
[ 3] 5901/tcp                   ALLOW IN    Anywhere                  
[ 4] 22/tcp (v6)                ALLOW IN    Anywhere (v6)             
[ 5] 5000:5999/tcp (v6)         ALLOW IN    Anywhere (v6)             
[ 6] 5901/tcp (v6)              ALLOW IN    Anywhere (v6) 
```

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been approved. -->

- [x] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Add a tag or create a release.

dav3r committed:
Deny any ports not needed for Guacamole to function properly

If assessors want to allow these ports, they must manually change these rules from "deny" to "allow".

dav3r committed:
Enable ufw and enforce firewall rules by default

It's preferable to have the default state of these instances be locked down with ufw enabled.

dav3r committed:
Bump version from 0.4.14+build.2 to 0.5.0

dav3r committed:
Bump version from 0.5.0 to 0.5.0-rc.1

jsf9k commented on src/ufw.yml:
```diff
@@ -6,9 +6,8 @@
-    # Do not actually enforce firewall rules.  The assessors will turn
-    # UFW on if they want it.
-    state: disabled
+    # Enable ufw and enforce firewall rules by default.
+    state: enabled
```
These lines can be removed.  `enabled` is [the default behavior](https://github.com/cisagov/ansible-role-ufw/blob/41e0944c9f2d17ce58e68812f1cc0386b6d32872/defaults/main.yml#L11-L13).

dav3r replied:
Good call- taken care of in https://github.com/cisagov/kali-packer/pull/92/commits/04b5eb8ad1f407989606bf2e2797790d4a7c0dfb.

jsf9k commented on src/ufw.yml:
```diff
@@ -23,4 +22,12 @@
       loop:
         - "22"       # ssh for Guacamole server (internal to assessment env)
         - "5901"     # vnc for Guacamole (internal to assessment env)
+    - name: Deny incoming traffic to specified TCP ports
+      # These rules may be manually modified by assessors as needed, but we
+      # want to default to a minimal set of allowed ports.
+      community.general.ufw:
+        proto: tcp
+        rule: deny
+        to_port: "{{ item }}"
+      loop:
         - 8000:8999  # Additional listeners, if necessary (external)
```
This section can be removed as well.  The default behavior for incoming traffic is `deny`.

dav3r replied:
In https://github.com/cisagov/cool-system-internal/issues/80, it sounds like @m1j09830 wants these `deny` rules to be explicitly defined, but I'll let him confirm this.  I'm happy to remove this if it's not needed.

m1j09830 replied:
@dav3r yes sir that is the idea. If it's explicitly defined then the teams will have a greater likelihood of identifying when something isn't working due to a firewall rule vs a technical issue on their end. For example if they are hosting payloads and unable to hit the payload site but don't see a firewall rule they may assume something is wrong with the hosting server and spend cycles going down an unnecessary rabbit hole. Just trying to prevent as much confusion upfront as possible.

jsf9k replied:
@dav3r - Please add a comment stating that those deny rules were explicitly requested in cisagov/cool-system-internal#80; otherwise, I will forget that and will be trying to remove them every time I touch this code.

dav3r replied:
Comment added in https://github.com/cisagov/kali-packer/pull/92/commits/789f695e129fa708b54ac84d85e2986de3060319 - please let me know if that works for you.

jsf9k replied:
The comment is great.  Are the ports 8000-8999 still relevant, or do those need to be changed to 5000-5999?

dav3r replied:
If I'm understanding @m1j09830 correctly, it sounds like those ports are still relevant here because they may be used by assessors on an as-needed basis.  @m1j09830, please correct me if that is wrong.

m1j09830 replied:
@dav3r we can go ahead and remove the deny rule since there are no default services running over those ports. The implicit deny on UFW will suffice. 

@jsf9k we will need to include the port range 5000-5999 as this range is used for internal comms between hosts. This should be in a default ALLOW state though since they are for internal comms only and have no risk of being exposed externally.

dav3r replied:
> This should be in a default ALLOW state though since they are for internal comms only and have no risk of being exposed externally.

I wouldn't say that there is **no risk** of those ports being exposed externally.  There is still the possibility of someone misconfiguring an environment and opening up some/all of those ports publicly, but the risk of that is fairly low with our current deployment process.

dav3r replied:
Please see https://github.com/cisagov/kali-packer/pull/92/commits/0a537c9fadba4e99d513c3e006e9d3e5dd835f57 and https://github.com/cisagov/kali-packer/pull/92/commits/1423aefd504df01947b3958d740cd31559f18c3b for the requested updates and let me know if anything looks awry now.

jsf9k replied:
Looks good to me!

jsf9k reviewed (commented):
A few comments for your consideration.

dav3r committed:
Remove ufw state variable 

We are no longer changing it from its default value, so it is no longer needed.

Co-Authored-By: Shane Frasier <maverick@maverickdolphin.com>

dav3r committed:
Add a comment explaining why we create "deny" ufw rules

Co-Authored-By: Shane Frasier <maverick@maverickdolphin.com>

dav3r committed:
Bump version from 0.5.0-rc.1 to 0.5.0-rc.2

dav3r committed:
Add TCP 5000-5999 to the list of allowed ports

These ports are needed to be in sync with the changes in https://github.com/cisagov/cool-assessment-terraform/pull/163.

dav3r committed:
After further discussion with Assessments, no "deny" rules are needed

dav3r committed:
Bump version from 0.5.0-rc.2 to 0.5.0-rc.3

jsf9k reviewed (approved):
The strongest of work, sir!

dav3r committed:
Bump version from 0.5.0-rc.3 to 0.5.0
